docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,186 @@
+
+```golang
+// StdSignature represents a sig
+type StdSignature struct {
+  PubKey        crypto.PubKey
+  Signature     []byte
+  PubKeyIndex   uint
+}
+```
+
+If a `StdSignature.PubKeyIndex` is 0, this means that the account's "master" PubKey (the `SubKeyAccount.PubKey`) should be used to verify the signature.  If the `StdSignature.PubKeyIndex` is `> 0`, it verifies the signature using the corresponding **1-indexed** SubKey in the `SubKeyAccount.SubKeys` slice.  Note, the SubKeys slice is 1-indexed rather than 0-indexed because the 0 index is used to refer to the master PubKey.  See pseudocode below:
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies that the SubKey has not been revoked and it is permitted to use the msg routes it is trying to sign over.
+
+```
+if acc.SubKeys[stdSig.PubKeyIndex - 1].Revoked {
+  return ErrNotPermitted
+}
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
+  return ErrNotPermitted
+}
+```
+
+When a transaction signed by a SubKey is accepted and has paid fees, we need to log the fees it has paid.  We increase the `SubKeyMetadata.DailyFeeUsed` field.
+
+If a new transaction from a subkey comes in and the `tx.Fee + subKeyMetadata.DailyFeeUsed > subKeyMetadata.DailyFeeAllowance`, then the transaction is rejected as the subkey is trying to exceed its daily fee allowance.
+
+But now we need a way to decrease the `DailyFeeUsed` field once transactions are past the 24 hour window.  To do this we us a time based iterator similar to the ones used in the governance and staking queues.
+
+We start by defining a new struct called a `DailyFeeSpend`, which is what will be inserted into the Queue.
+
+```golang
+type DailyFeeSpend struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  FeeSpent             sdk.Coins
+}
+```
+
+When a transaction using a SubKey has paid its fee, we insert a new entry into this queue's store.
+- The key is `sdk.FormatTimeBytes(ctx.BlockTime.Add(time.Day * 1)`
+- The value is the Amino marshalled `DailyFeeSpend`
+
+Finally in the EndBlocker we iterate over the all the DailyFeeSpends in the queue which are expired (more than a day old).  We prune them from state, and deduct their FeeSpent amount from the corresponding SubKeyMetadata.
+
+```
+store := ctx.Store(spentFeeQueueStoreKey)
+iterator := sdk.Iterator(nil, sdk.FormatTimeBytes(ctx.BlockTime)) // pseudocode abstract this as a slice of DailyFeeSpend
+
+for _, dailyFeeSpend := range iterator {
+  acc := accountKeeper.GetAccount(dailyFeeSpend.Address)
+  acc.SubKeys[dailyFeeSpend.SubKeyIndex - 1].DailyFeeUsed -= dailyFeeSpend.FeeSpent
+  store.Delete(dailyFeeSpend)
+}
+```
+
+## Msgs
+
+In order to add, revoke, and update SubKeys, we would need to add some new `Msg` types to the auth module.
```

Comment:
shouldn't we add a message to also update the `PermissionedRoutes` of a `SubKey` ?

Reply:
maybe merge it with `MsgUpdateSubKeyAllowance` to just have a `MsgUpdateSubKey` for better UX. I personally wouldn't like to submit 2 transactions to update the routes and the allowance

Reply:
You can just create a new subkey with the same `crypto.PubKey`.